src/debug.rs

```diff
@@ -0,0 +1,352 @@
+{
+    fn trace_block_by_number(
+        &self,
+        _block: BlockNumber,
+        _options: Option<TracerConfig>,
+    ) -> BoxFuture<Result<Vec<ResultDebugCall>>> {
+        todo!()
+    }
+
+    fn trace_block_by_hash(
+        &self,
+        _hash: H256,
+        _options: Option<TracerConfig>,
+    ) -> BoxFuture<Result<Vec<ResultDebugCall>>> {
+        todo!()
+    }
+
+    /// Trace execution of a transaction.
+    fn trace_call(
+        &self,
+        request: CallRequest,
+        block: Option<BlockId>,
+        options: Option<TracerConfig>,
+    ) -> BoxFuture<Result<DebugCall>> {
+        let only_top = options.is_some_and(|o| o.tracer_config.only_top_call);
+        let inner = Arc::clone(&self.node);
+        Box::pin(async move {
+            if block.is_some() && !matches!(block, Some(BlockId::Number(BlockNumber::Latest))) {
+                return Err(jsonrpc_core::Error::invalid_params(
+                    "tracing only supported at `latest` block",
+                ));
+            }
+
+            let inner = inner
+                .read()
+                .map_err(|_| into_jsrpc_error(Web3Error::InternalError))?;
+
+            let mut l2_tx = match L2Tx::from_request(request.into(), MAX_TX_SIZE) {
+                Ok(tx) => tx,
+                Err(e) => {
+                    let error = Web3Error::SerializationError(e);
+                    return Err(into_jsrpc_error(error));
+                }
+            };
+
+            let execution_mode = TxExecutionMode::EthCall;
+            let storage = StorageView::new(&inner.fork_storage).to_rc_ptr();
+
+            let bootloader_code = inner.system_contracts.contacts_for_l2_call();
+
+            // init vm
+            let (l1_batch_env, _block_context) = inner.create_l1_batch_env(storage.clone());
+            let system_env = inner.create_system_env(bootloader_code.clone(), execution_mode);
+            let mut vm = Vm::new(l1_batch_env, system_env, storage, HistoryDisabled);
+
+            // We must inject *some* signature (otherwise bootloader code fails to generate hash).
+            if l2_tx.common_data.signature.is_empty() {
+                l2_tx.common_data.signature =
+                    PackedEthSignature::default().serialize_packed().into();
+            }
+
+            let tx: Transaction = l2_tx.clone().into();
+            vm.push_transaction(tx);
+
+            let call_tracer_result = Arc::new(OnceCell::default());
+            let tracer = CallTracer::new(call_tracer_result.clone(), HistoryDisabled);
+            let tx_result = vm.inspect(vec![Box::new(tracer)], vm::VmExecutionMode::OneTx);
+
+            let call_traces = if only_top {
+                vec![]
+            } else {
+                Arc::try_unwrap(call_tracer_result)
+                    .unwrap()
+                    .take()
+                    .unwrap_or_default()
+            };
+
+            let calltype = if l2_tx.recipient_account() == CONTRACT_DEPLOYER_ADDRESS {
```

Comment:
Note: I _believe_ your implementation is the correct `r#type`, but it does differ from `zksync_core` as that will always return a `DebugCallType:Create` based on the following logic: https://github.com/matter-labs/zksync-era/blob/main/core/lib/types/src/vm_trace.rs#L110

Probably worth creating an Issue/PR in `zksync_era` to address this discrepancy